Add Storage bat tests

markdryan opened the pull request:
bat: Add storage bat tests
    
This PR adds the following storage BAT tests
    
- Test bootable volumes are created and deleted correctly
- Check bootable volumes of stopped instances are in-use
- Check that in-use volumes cannot be deleted
- Check that the rootfs volumes cannot be detached
- Check that volumes can be added, listed and deleted
- Check attaching volumes become available when deleting their instance.
- Check that a deleting a detaching volume fails
- Check that we can attach and detach volumes
- Check that we cannot attach to a running container
- Check that we cannot attach a volume to two instances
- Check we can attach a volume to a stopped instance
    
As we now have multiple BAT tests that create and delete instances in the same tenant the BAT tests should be run with test-cases, rather than go test, as the latter runs tests for separate packages concurrently, which is not what we currently want.

This PR also splits the bat.go file into multiple go files to make it easier to navigate.  Finally, it updates the ciao-cli command to take advantage of the new functions in the ciao-cli template language. 
    
Fixes #682

markdryan commented:
Not all these new BAT tests are currently passing.  This is due to the following bugs in controller:

- #875 (This doesn't cause a BAT test failure but needs to be fixed to manually test storage)
- #872
- #871
- #870
- #867
- #869  (This doesn't cause a BAT test failure)

kaccardi commented on _release/bat/storage_bat/storage_bat_test.go:
```diff
@@ -0,0 +1,623 @@
+
+	err = bat.WaitForVolumeStatus(ctx, "", volumeID, "available")
+	if err != nil {
+		t.Errorf("Timed out waiting for volume %s to become availabled : %v",
```
is "availabled" a typo, or a new verb describing the act of becoming available?

markdryan commented:
I update the commit to fix the spelling mistake.  I also changed the behaviour of the AttachedToRunningContainer test to reflect the expected behaviour and fixed some issues with the use of defer to clean up volumes in some of the tests.  All tests are now passing locally for me with

#885 , #886, #887

markdryan commented:
@tpepper I've rebased.  All tests are passing on master now.  We can merge.

markdryan commented:
Rebased.  Tests still pass.

@tpepper Could you review and merge?
